### DISCUSSION TITLE
What remark and rehype plugins are people using?

### DISCUSSION THREAD

**Discussion by pngwn** [pngwn/MDsveX#354] [General]
As mdvsex will be moving away from the unified ecosystem in 1.0, I'd love to know what plugins people are using currently so that I can ensure nothing is missing when this change is made.

If you comment with the plugin, what you're using it for, and a link to the source, that would be great.

If you are using a custom plugin of your own making, I'd also love to hear about that too.

And if there is a plugin that you _wished_ exist but doesn't, that would also be very useful.

**Comment by Madd0g** [comment]
I'm using these public plugins:

* rehype-slug
* remark-wiki-link
* remark-toc
* remark-sectionize
* remark-html

I wrote my own plugins to:

* remark: do special things with sections (after `remark-sectionize` creates them)
  * like giving sections html props (assigning a slot prop to a `section` to teleport it into a named slot is amazing!)
* remark: automatically injecting the TOC heading for the `remark-toc` plugin
* remark: parse some things from the AST and inject into frontmatter
* rehype: parse some things from the HTML and inject into frontmatter

I've actually just started getting into it and creating my own plugins, this change scares me...

**Reply by pngwn** [reply]
Nothing to be scared about! they'll work in a similar way to remark/ rehype plugins but the API will be a good deal simpler. There is some additional context for these changes [in this discussion](https://github.com/pngwn/MDsveX/discussions/259). Other changes coming [are discussed here](https://github.com/pngwn/MDsveX/discussions/293).

The current API for rehype/remark plugins is very complex and not something I want mdsvex users dealing with , so I don't personally see an alternative due to the direction the unified ecosystem has gone in.

**Reply by Masstronaut** [reply]
@Madd0g How did you set up wiki-links? It's been on my to-do list and I haven't gotten to it yet; I store page titles & slugs in the frontmatter, and don't know how to get the list of slugs early enough to provide them to the wiki-links plugin.

**Reply by Madd0g** [reply]
@Masstronaut - my setup is not smart, anything wrapped in `[[ ]]` would just turn into a link, it doesn't check if the page exists yet.

```js
    remarkPlugins: [
      [
	      wikiLinkPlugin,
	      {
		      hrefTemplate(permalink) {
			      return permalink;
		      },
		      wikiLinkClassName: 'internal wikilink'
	      }
      ],
```

EDIT: actually I can see you can pass a list of permalinks to it:


> options.permalinks [String]: An array of permalinks that should be considered existing pages. If a wiki link is parsed and its permalink matches one of these permalinks, node.data.exists will be true.

**Reply by Masstronaut** [reply]
Yeah I've looked at the plugin, but in the context of MDsveX I haven't figured out a clean way to provide the permalinks list since I don't have it until after the frontmatter for every note has been parsed. @pngwn any ideas?

**Reply by Madd0g** [reply]
Yeah, that's definitely a problem, I wish there was a good solution for this problem. I was going to try to let svelte handle that part, by using the [custom components](https://mdsvex.com/docs#custom-components) feature for wikilinks.

Still not sure if that's the best idea, but it's the only one I currently have, inspired by #100

**Reply by Masstronaut** [reply]
Okay, that doesn't seem too bad. From the [remark-wiki-link readme](https://github.com/landakram/remark-wiki-link) it looks like it adds `class="internal new"` which should make it easy to find wikilinks vs normal ones. I'll give it a go in the next week or two and report back.

**Reply by wighawag** [reply]
@Masstronaut I am interested in the wikilinks setup
So far I got it working by using remark-wiki-link@0.0.4 I guess that's the same for you ?
And I use the same config as you above, but to ensure the `new` class is added only to non-existing link, I create a permalink -> pagePath map and use it in the hrefTemplate. The shortest path take precedence. Not sure if this is the way forward and happy to hear how you deal with it. thanks!

**Reply by webel** [reply]
@Madd0g how'd you get remark-toc to work? So far nothing seems to happen after adding the plugin :(

**Comment by babichjacob** [comment]
For a library documentation site I've used:
* [`remark-github`](https://github.com/remarkjs/remark-github) to include references to usernames like @babichjacob just by writing that text.
* [`rehype-external-links`](https://github.com/rehypejs/rehype-external-links) to include basic security like `rel="nofollow"` on external links and adding screenreader-only text "opens in a new window". I'd also imagine this can allow me to add an arrow-upright icon to external links or otherwise style them differently but I haven't tried that yet.
* [`rehype-slug`](https://github.com/rehypejs/rehype-slug) + [`rehype-autolink-headings`](https://github.com/rehypejs/rehype-autolink-headings): jump to headings with `#hashes` in the URL and also click on headings to update the URL with their hash.
* [`remark-abbr`](https://github.com/zestedesavoir/zmarkdown/tree/master/packages/remark-abbr) for abbreviations

I am so far using these for my school notes site / app:
* [`remark-directive`](https://github.com/remarkjs/remark-directive) for any, many custom container directives, leaf blocks, and inline directives, like in this snippet:
  ```md
    :::::::::annotatedslides
    ::::::annotatedslide
    ::embed[./_Lecture-W4-Criminal-Law/Page-1.pdf]
    :::annotation
    :term[Criminal law] is :definition[the foundation of the criminal justice system].
    :::
    ::::::
    ... and so on for the other pages
    :::::::::
    ``` 
   or this snippet:
   ```md
   # Homework (this is made up)
   :::qna
   ::question[What is the slope of $y = 2x + 5$?]
   ::answer[The slope is the coefficient of $x$ in linear equations, so here that's :finalanswer[$2$].]
   :::
   ... and so on for the other (usually homework) questions
   ```
    
  which works like:
  * `:term[Word]` and `:definition[definition of the word]` style these in an exciting way. I also expect to use some other kind of processing to hunt down all these terms and definitions across my class notes to generate flash cards for them. It'd be nice to have args so that `:definition[for=Word][definition of the word]` or any other syntax means that there isn't any issues with ambiguity. I might want to do something similar for questions and answers.
  * `::embed[https://www.youtube.com/watch?v=S3j1fLzC8_E]` to embed a YouTube video with their nice look. If I wanted to embed other things I could use [`remark-oembed`](https://github.com/agentofuser/remark-oembed) as a reference, which I didn't directly use so I could have customizability + overload the embed directive with support for e.x. `::embed[./slides-page-1.pdf]` or `::embed[./assigned-reading.md]` or `::embed[https://mdsvex.com/]` and it automatically figure out what to do. I also think this is a reasonable solution for https://github.com/pngwn/MDsveX/discussions/299 when you're willing to refactor repeated or referenced content into its own file.
  * `annotatedslides`, `annotation`, `question`, etc, are mainly so things can be styled as I would like (they create `div`s or `span`s with custom classes).

* [This custom remark plugin](https://github.com/pngwn/MDsveX/discussions/246#discussioncomment-732230) for automatically importing assets for image directives (`
![Pipelines](./_pipeline-diagram.jpg)`) and the previously mentioned `embed`

* [`remark-gfm`](https://github.com/remarkjs/remark-gfm) for automatically converting bare links in text to links and for okay table support. I also might use `- [x] todo` check lists eventually.
* [`remark-math` + `rehype-katex`](https://github.com/remarkjs/remark-math) for math in common LaTeX syntax. I saw on another discussion post that custom syntax blocks can be used to avoid issues with `{}` as an alternative to these plugins but I've been fine with these so long as I define variables that it *looks* like they use (e.x. `$x^{y}$` renders fine after adding a `script` block with `let y;`)
* The previously mentioned [`rehype-slug`](https://github.com/rehypejs/rehype-slug) + [`rehype-autolink-headings`](https://github.com/rehypejs/rehype-autolink-headings)


Other things I want to use in general, but haven't yet, are:
* [This custom remark plugin](https://github.com/pngwn/MDsveX/discussions/268#discussioncomment-831882) for finding headers and adding them to frontmatter for custom table of content generation (especially across pages)
* [`remark-reading-time`](https://github.com/mattjennings/remark-reading-time)

**Reply by Madd0g** [reply]
@babichjacob are you using `remark-directive` in mdsvex currently? Is there a version that works with the current remark version? I couldn't get it to work in mdsvex

**Reply by babichjacob** [reply]
Yes, my lockfile has it at `2.0.0`. I also needed to put `remark-parse` and `remark-frontmatter` before it because `unified` ecosystem.

**Reply by Madd0g** [reply]
I added those 2 libs before `remark-directive` and it works now! Sweet!
Thanks

**Reply by amr3k** [reply]
May I ask you how did you get [remark-reading-time](https://github.com/mattjennings/remark-reading-time) working?

**Reply by lukens** [reply]
I added `remark-parse`, and that got `remark-directive` working, but broke tables and strikethrough and other extended syntax that had worked previously. Adding `remark-gfm` (GitHub Flavoured Markdown) got those working again.

**Reply by lukens** [reply]
Thinking about this some more, I think adding `remark-parse` as a plugin is effectively junking all the parsing that was done up until that point, and re-parsing the markdown file from scratch. As such it's no longer being parsed by the mdsvex parser, and so will, presumably, be losing much of the mdsvex goodness.

@pngwn - Can you confirm the above, and so is this the wrong approach to take? Is there is a _right_ approach to use `remark-directive`, or are we out of luck?

Having skimmed some other discussions/issues, is this maybe related to #236? Does [this comment](https://github.com/pngwn/MDsveX/issues/76#issuecomment-674267071) suggest the use of `{...}` in the directives syntax will always be a problem?

I am also pondering why I am trying to use directives at all, when I could just use Svelte components. I think my main reasons are that I prefer the more markdown-y syntax of directives, and also like that directives have access to the syntax tree at parsing time, so can modify child or parent elements; whereas, as far as I can figure out, Svelte components would only be able to do this via the DOM after rendering (which is likely to lead to a flash of some kind whilst stuff is moved about).

**Reply by webel** [reply]
@lukens did you find a way around junking the mdsvex parser? I'm also inclined to use directives for our docs, but like the idea of svx for some components...

**Reply by lukens** [reply]
> @lukens did you find a way around junking the mdsvex parser? I'm also inclined to use directives for our docs, but like the idea of svx for some components...

Hi, sorry I missed this comment before. I've not found a solution for this in mdsvex. Dropping plugin support also makes me a little worried about the direction of the mdsvex project in general, so I have had a brief look at alternatives. In particular I've toyed with moving my docs site over to [KitDocs](https://kit-docs.svelteness.dev/docs/production/deploying#static-builds), as it already implements a lot of features I want. It does however also have some aspects I'm not keen on (it's also quite broken currently, due to the SvelteKit routing updates), so I've also had a quick experiment with just using the Markdown parser from KitDocs. That showed some promise, but is also pretty entangled with a lot of the other features of KitDocs. I may end up just using a pure Markdown parser, and forgetting Svelte support in Markdown docs, or might look at adding Svelte support myself (possibly starting with a fork of KitDocs, if the licence looks suitably permissive).

**Reply by ahopkins** [reply]
So, I tried the idea of skipping directives and just using a Svelte component, but then I end up with un-parsed markdown. :unamused: 

There must be a solution out there without having to make an ugly nesting of components and rendering.

**Reply by janosh** [reply]
Curious if anyone has news to share about using [remark-directive](https://github.com/remarkjs/remark-directive) with `mdsvex`. `remark-directive` looks like it offers a lot of extensibility.

**Reply by magicpalmtree** [reply]
> For a library documentation site I've used:
> 
> * [`remark-github`](https://github.com/remarkjs/remark-github) to include references to usernames like @babichjacob just by writing that text.
> * [`rehype-external-links`](https://github.com/rehypejs/rehype-external-links) to include basic security like `rel="nofollow"` on external links and adding screenreader-only text "opens in a new window". I'd also imagine this can allow me to add an arrow-upright icon to external links or otherwise style them differently but I haven't tried that yet.
> * [`rehype-slug`](https://github.com/rehypejs/rehype-slug) + [`rehype-autolink-headings`](https://github.com/rehypejs/rehype-autolink-headings): jump to headings with `#hashes` in the URL and also click on headings to update the URL with their hash.
> * [`remark-abbr`](https://github.com/zestedesavoir/zmarkdown/tree/master/packages/remark-abbr) for abbreviations
> 
> I am so far using these for my school notes site / app:
> 
> * [`remark-directive`](https://github.com/remarkjs/remark-directive) for any, many custom container directives, leaf blocks, and inline directives, like in this snippet:
>   ```lisp
>     :::::::::annotatedslides
>     ::::::annotatedslide
>     ::embed[./_Lecture-W4-Criminal-Law/Page-1.pdf]
>     :::annotation
>     :term[Criminal law] is :definition[the foundation of the criminal justice system].
>     :::
>     ::::::
>     ... and so on for the other pages
>     :::::::::
>   ```
>   
>   
>       
>         
>       
>   
>         
>       
>   
>       
>     
>   or this snippet:
>   ```lisp
>   # Homework (this is made up)
>   :::qna
>   ::question[What is the slope of $y = 2x + 5$?]
>   ::answer[The slope is the coefficient of $x$ in linear equations, so here that's :finalanswer[$2$].]
>   :::
>   ... and so on for the other (usually homework) questions
>   ```
>   
>   
>       
>         
>       
>   
>         
>       
>   
>       
>     
>   which works like:
>   
>   * `:term[Word]` and `:definition[definition of the word]` style these in an exciting way. I also expect to use some other kind of processing to hunt down all these terms and definitions across my class notes to generate flash cards for them. It'd be nice to have args so that `:definition[for=Word][definition of the word]` or any other syntax means that there isn't any issues with ambiguity. I might want to do something similar for questions and answers.
>   * `::embed[https://www.youtube.com/watch?v=S3j1fLzC8_E]` to embed a YouTube video with their nice look. If I wanted to embed other things I could use [`remark-oembed`](https://github.com/agentofuser/remark-oembed) as a reference, which I didn't directly use so I could have customizability + overload the embed directive with support for e.x. `::embed[./slides-page-1.pdf]` or `::embed[./assigned-reading.md]` or `::embed[https://mdsvex.com/]` and it automatically figure out what to do. I also think this is a reasonable solution for [Easier import of page content #299](https://github.com/pngwn/MDsveX/discussions/299) when you're willing to refactor repeated or referenced content into its own file.
>   * `annotatedslides`, `annotation`, `question`, etc, are mainly so things can be styled as I would like (they create `div`s or `span`s with custom classes).
> * [This custom remark plugin](https://github.com/pngwn/MDsveX/discussions/246#discussioncomment-732230) for automatically importing assets for image directives (` ![Pipelines](./_pipeline-diagram.jpg)`) and the previously mentioned `embed`
> * [`remark-gfm`](https://github.com/remarkjs/remark-gfm) for automatically converting bare links in text to links and for okay table support. I also might use `- [x] todo` check lists eventually.
> * [`remark-math` + `rehype-katex`](https://github.com/remarkjs/remark-math) for math in common LaTeX syntax. I saw on another discussion post that custom syntax blocks can be used to avoid issues with `{}` as an alternative to these plugins but I've been fine with these so long as I define variables that it _looks_ like they use (e.x. `$x^{y}$` renders fine after adding a `script` block with `let y;`)
> * The previously mentioned [`rehype-slug`](https://github.com/rehypejs/rehype-slug) + [`rehype-autolink-headings`](https://github.com/rehypejs/rehype-autolink-headings)
> 
> Other things I want to use in general, but haven't yet, are:
> 
> * [This custom remark plugin](https://github.com/pngwn/MDsveX/discussions/268#discussioncomment-831882) for finding headers and adding them to frontmatter for custom table of content generation (especially across pages)
> * [`remark-reading-time`](https://github.com/mattjennings/remark-reading-time)

@babichjacob did you checked footnotes inside table with remark-gfm?
currently I am using remark-gfm to build our docs and it's almost perfect that converts all markdowns to html except for footnotes inside table.
here is example input:
```
Before table [^t0]

| h1 [^h1] | h2 [^h2] | h3 [^h3] |
| --- | --- | --- |
| 1 [^c1] | 2 [^c2] | 3 [^c3] | 
| 4 [^c4] | 5 [^c5] | 6 [^c6] |
| 7 [^c7] | 8 [^c8] | 9 [^c9] |

After table[^t1]

[^t0]: t0-desc
[^h1]: h1-desc
[^h2]: h2-desc
[^h3]: h3-desc
[^c8]: c8-desc
[^c9]: c9-desc
[^c1]: c1-desc
[^t1]: t1-desc
[^c2]: c2-desc
[^c3]: c3-desc
[^c4]: c4-desc
[^c5]: c5-desc
[^c6]: c6-desc
[^c7]: c7-desc
```

there is no issue on remark-gfm itself, I already checked and discussed with remark dev team.
```
import {unified} from 'unified'
import remarkParse from 'remark-parse'
import remarkGfm from 'remark-gfm'
import remarkRehype from 'remark-rehype'
import remarkStringify from 'rehype-stringify'

const file = await unified()
  .use(remarkParse)
  .use(remarkGfm)
  .use(remarkRehype)
  .use(remarkStringify)
  .process(
    `Before table [^t0]

| h1 [^h1] | h2 [^h2] | h3 [^h3] |
| --- | --- | --- |
| 1 [^c1] | 2 [^c2] | 3 [^c3] | 
| 4 [^c4] | 5 [^c5] | 6 [^c6] |
| 7 [^c7] | 8 [^c8] | 9 [^c9] |

After table[^t1]

[^t0]: t0-desc
[^h1]: h1-desc
[^h2]: h2-desc
[^h3]: h3-desc
[^c8]: c8-desc
[^c9]: c9-desc
[^c1]: c1-desc
[^t1]: t1-desc
[^c2]: c2-desc
[^c3]: c3-desc
[^c4]: c4-desc
[^c5]: c5-desc
[^c6]: c6-desc
[^c7]: c7-desc
`
  )

console.log(String(file))
```
the code above is working fine.

but why not working when its combined with mdsvex as plugin?

**Comment by Masstronaut** [comment]
footnotes, slug, headings

I find that MDsveX handles markdown tables already, but that (and GFM in general) are important features for me. 

I also use a handful of my own custom plugins to do things like:

- Calculate time to read a post and add the result to my frontmatter metadata
- find custom content embedded in the markdown (similar to @babichjacob ) such as flashcards and add them to frontmatter
  - Remove some special identifiers that are related to this content (in my case `:-:` and `:_:`) so that the text renders cleanly
 
Generally I'd say that @babichjacob seems to be doing a lot of the things I'm either already doing or planning to do in the next couple months, so +1 to all their use cases.

**Reply by babichjacob** [reply]
Can you share any amount of your content-hunting plugin(s) code? I can't decide whether to try to tackle the problem with normal JavaScript parsing text or with remark

**Reply by Masstronaut** [reply]
Sure, code snippet below. I haven't yet added support for "rich" content (images, links, em/strong, code) in my flashcards, so this will not cover those cases. If you want to support those sorts of things in your flashcards, I would recommend the remark route. Parsing them by yourself is not worth the hassle IMO. If you don't intend to support those... I recommend you reconsider 😂

```js
import visit from 'unist-util-visit';

export function mdFlashcardGrabber() {
	return transformer;
}

// finds all flashcards with prompt and answer separated by :-:
// Puts the flashcards into the metadata object for later use
const transformer = (ast, vFile) => {
	// Collect all the flashcards from this md file
	const cards = [];
	visit(ast, "paragraph", visitor);
	// store the flashcards where they can be accessed
	// vFile.data.fm corresponds to:
	// import.meta.globEager('path/to/files/*.md')[index].metadata
	vFile.data.fm.flashcards = [...cards];

	function visitor(node) {
		// Store the paragraph's collected text in the results.
		// parts before it is identified as a flashcard go in "prompt"
		// parts after it's identified as a flashcard go in "answer"
		const results = { prompt: "", answer: "", flashcard: false };

		// Walk the nodes in this paragraph to determine if it's a flashcard
		// Build up the flashcard object as we go so that if it is one
		// it doesn't need to be traversed again.
		function traverseChildren(node, results) {
			// Content text only exists in "text" nodes.
			// For any other kind we can just check its children.
			if (node.type === "text") {
				if (!node.value) return;
				// If the flashcard indicator was already found,
				// any text after it is part of the answer
				if (results.flashcard) {
					results.answer += node.value;
					return;
				}
				// Split the text by the flashcard indicator.
				const text = node.value.split(':-:');
				if (text.length === 1) {
					// no flashcard indicator here
					results.prompt += text[0];

					if (node.value.includes(":_:")) {
						// this node contains a multi-paragraph flashcard indicator
						// For now these aren't implemented, so remove the indicator
						// @@TODO: Support multi-line flashcards
						console.warn("multi-line flashcard detected. Support for them has not been implemented yet.");
						node.value = node.value.replace(":_:", "\n");
					}
				}
				else {
					// Remove my flashcard indicators from the rendered text
					node.value = node.value.replace(":-:", " ");
					if (text.length === 2) {
						// Found a single flashcard indicator - this is a flashcard
						results.prompt += text[0];
						results.flashcard = true;
						results.answer += text[1];
					}
					else {
						// unexpected number of flashcard indicators (>1 per paragraph)
						// assume it's meant to be 1
						results.prompt += text[0];
						results.flashcard = true;
						text = [, ...text]; // remove the first element of the array
						text.forEach(s => results.answer += s);
					}
				}
			}
			else {
				if (node.children) {
					node.children.forEach(child => traverseChildren(child, results));
				}
			}
		}
		traverseChildren(node, results);
		if (results.flashcard) {
			cards.push({ prompt: results.prompt, answer: results.answer });
		}
	}
}

```

**Comment by TheOnlyTails** [comment]
Over at the Files website (https://files.community/) we use:
- `@fec/remark-a11y-emoji` - makes emojis more accessible.
- `remark-slug` - adds slug IDs to headings: `# Title` -> `<h1 id="title">Title</h1>`
- `remark-gfm` - adds support for [some GFM features](https://www.npmjs.com/package/remark-gfm#use)
- `remark-github` - allows for quick linking to GitHub https://www.npmjs.com/package/remark-github#use

`svelte.config.js` for reference: https://github.com/files-community/Website/blob/main/svelte.config.js

On my personal site (https://theonlytails.com/), I use:
- All of the previously mentioned plugins.
- `remark-gemoji`: adds support for `:emoji_name:`-style emojis.
- `remark-hint`: allows to add warnings, errors and hints.
- `rehype-external-links`: turns all the links in the markdown page into external links.

`svelte.config.js` for reference: https://github.com/theonlytails/theonlytails.com/blob/main/svelte.config.js

**Comment by bwklein** [comment]
I've been looking, but haven't found it yet.  But I would like to have section numbers added automatically to the headings.

```
# Example 

Some Text ...

## Subheading

More Text ...
```

becomes

```
<h1>1.0 Example</h1>

Some Text ...

<h2>1.1 Subheading</h2>

More Text ...
```

**Reply by lukens** [reply]
It should be fairly easy to write a plugin to do this, however, you can also achieve this with css (and, arguably it is a presentation consideration, so is maybe best handled by css):

https://www.sqlpac.com/en/documents/html-css-auto-numbering-elements-counters.html

**Reply by janosh** [reply]
Haven't looked into a plugin solution but like @lukens said, this sort of CSS does the trick for me.

```css
/* auto-number headings, import this CSS in a .svx file to apply */
main > h2 {
  counter-increment: h2;
  counter-reset: h3, h4;
}
main > h2::before {
  content: counter(h2);
  margin-right: 10pt;
}
main > h3 {
  counter-increment: h3;
  counter-reset: h4;
}
main > h3::before {
  content: counter(h2) '.' counter(h3);
  margin-right: 10pt;
}
main > h4::before {
  content: counter(h2) '.' counter(h3) '.' counter(h4);
  margin-right: 10pt;
}
```

The only potential downside is that since the numbers are added in `::before` pseudo-elements and not the heading node itself, the numbers won't show up in a table of contents. But I can also see that being desired behavior in some cases.

**Comment by rchrdnsh** [comment]
`rehype-add-classes` to, you know...add classes :-)

that plugin is also not maintained at all and has a dependency with severe vulnerabilities.

**Comment by lukens** [comment]
I have a few custom plugins.

- The first takes the first `<h1>` element, removes it from the html, and adds it to the metadata as `title` (unless a title is already provided).

- The second rewrites relative links to `.md` files, so that the links work on the routes in my site.

- I have one that provides nicer markup for quotations with attribution.

I also use [the plugin for making sure relative images work](https://www.npmjs.com/package/mdsvex-relative-images), as well a `remark-driectives` based one for callouts.

**Comment by magicpalmtree** [comment]
P.S. consider these two codes below.
```
import {unified} from 'unified'
import remarkParse from 'remark-parse'
import remarkGfm from 'remark-gfm'
import remarkRehype from 'remark-rehype'
import remarkStringify from 'rehype-stringify'

const file = await unified()
  .use(remarkParse)
  .use(remarkGfm)
  .use(remarkRehype)
  .use(remarkStringify)
  .process(
    `Before table [^t0]

| h1 [^h1] | h2 [^h2] | h3 [^h3] |
| --- | --- | --- |
| 1 [^c1] | 2 [^c2] | 3 [^c3] | 
| 4 [^c4] | 5 [^c5] | 6 [^c6] |
| 7 [^c7] | 8 [^c8] | 9 [^c9] |

After table[^t1]

[^t0]: t0-desc
[^h1]: h1-desc
[^h2]: h2-desc
[^h3]: h3-desc
[^c8]: c8-desc
[^c9]: c9-desc
[^c1]: c1-desc
[^t1]: t1-desc
[^c2]: c2-desc
[^c3]: c3-desc
[^c4]: c4-desc
[^c5]: c5-desc
[^c6]: c6-desc
[^c7]: c7-desc
`
  )

console.log(String(file))
```
```
import { mdsvex, compile } from 'mdsvex';
import remarkParse from 'remark-parse'
import remarkGfm from 'remark-gfm'

const file = await compile(
    `Before table [^t0]

| h1 [^h1] | h2 [^h2] | h3 [^h3] |
| --- | --- | --- |
| 1 [^c1] | 2 [^c2] | 3 [^c3] | 
| 4 [^c4] | 5 [^c5] | 6 [^c6] |
| 7 [^c7] | 8 [^c8] | 9 [^c9] |

After table[^t1]

[^t0]: t0-desc
[^h1]: h1-desc
[^h2]: h2-desc
[^h3]: h3-desc
[^c8]: c8-desc
[^c9]: c9-desc
[^c1]: c1-desc
[^t1]: t1-desc
[^c2]: c2-desc
[^c3]: c3-desc
[^c4]: c4-desc
[^c5]: c5-desc
[^c6]: c6-desc
[^c7]: c7-desc
`, {
  remarkPlugins: [remarkGfm, remarkParse]
}
  )

console.log(String(file.code))
```
the first one is working fine, but the latter one prints the footnotes in reverse order.
is this a mdsvex issue?

**Comment by babakfp** [comment]
Hi

Thank you for MDSveX.

These are all of the packages I use for markdown:

```
shiki
remark-unwrap-images
rehype-slug
rehype-external-links
rehype-autolink-headings
hastscript
```

**Comment by doppelganger9** [comment]
Hi,

Thank you for mdsvex!

In my blog, I use these remark plugins:

- `remark-abbr` : adds support for footnote-like abbreviations
- `remark-twemoji` : transforms emojis into [twemojis](https://twemoji.twitter.com/) for accessibility and to ensure same rendering on all OS browsers.
- `@akebifiky/remark-simple-plantuml` : to embed PlantUML diagrams.
- `@remark-embedder/core` and 
- `@remark-embedder/transformer-oembed` to transform URLs to auto-embed links as YouTube videos, GIFs, Tweets (Xs?), etc.

A custom Remark function to add word count to frontmatter:

```ts
function addWordCountToFrontmatterData({
  attribute = "wordCount",
} = {}) {
  return function (info, file) {
    let text = "";

    visit(info, ["text", "code"], (node) => {
      text += node.value;
    });
      
    file.data.fm[attribute] = text.split(/\s+/).length;
  };
}
```

And these rehype plugins:

- `rehype-urls` + custom function : adds `rel` and target to `<a>` elements
- `rehype-slug` : to add slug to `<h1>`-`<h6>` elements
- `rehype-autolink-headings` : to prepend slugged `<h1>`-`<h6>` elements with an `<a>`
- `rehype-add-classes` : to add `list` class to `ul,ol`
- `rehype-figure` : to wrap images into `figure` + `figcaption`

For context, see my blog's repository : https://github.com/doppelganger9/blog/blob/master/mdsvex.config.js

(It might not be optimal or 100% correctly configured)

**Reply by pdenapo** [reply]
The idea of "moving away from the unified ecosystem" does not seem to be a good one. Over time, this means that the community will split, and the users will find less useful resources and examples that do not work. Please refrain from doing that!

Answering your question, I am using rehypeSlug, rehypeAutolinkHeadings and rehypeKatex for my math blog.
Which is based on 

https://github.com/josh-collinsworth/sveltekit-blog-starter

rehypeKatex adds support for rendering equations in Latex, which is essential for my use-case.
